Use request context service instead of container params if no request is available

stollr opened the pull request:
This change fixes this issue (described in #7280).

| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | no
| Deprecations? | yes
| Fixed tickets | fixes #7280
| License | MIT

#### What's in this PR?

Changed the way how the default host and scheme is fetched in the WebspaceManager and DumpSitemapCommand. Instead of directly accessing the container params `router.request_context.host` and `router.request_context.scheme` the information is fetched via the `RequestContext` service of Symfony.

#### Why?

The webspace manager, which is configured in the core bundle, relied on the request context container parameters `router.request_context.host` and `router.request_context.scheme` as fallback when no actual request object is available. This might happen during cache warming in a functional test.

But if the request context is configured by Symfony's `framework.router.default_uri` setting like it is described in their [docs](https://symfony.com/doc/current/routing.html#generating-urls-in-commands), the mentioned container parameters still uses the defaults (localhost and http). There is a issue which points this out https://github.com/symfony/symfony/issues/53919.

This could have lead to wrong url generation in a functional test (kernel test case) when the cache is warmed.

mamazu commented:
Yes, I already ran into a similar issue when trying to move the Webspace config to Symfony config parsing. However I would suggest leaving the constuctor as is and just report that passing those arguments is deprecated. We can just have a function that sets the service into the class and have it fall back to the old implemenatation if it isn't there. Then we can clean it up in the 3.0 release with a small bc break.

stollr commented:
As far as I see there is no BC break for 2.x in my implementation. But adding a setter for the service will introduce one more BC break to 3.0.

alexander-schranz commented:
Think a none bc break here would be change the service definition to an expression:

```diff
-            <argument>%router.request_context.scheme%</argument>
-            <argument>%router.request_context.host%</argument>
+            <argument type="expression">service('router.request_context').getScheme()</argument>
+            <argument type="expression">service('router.request_context').getHost()</argument>
```

stollr commented:
@alexander-schranz The service definition file does not have a BC promise, does it?

In the end, the constructor already fetches the values from the request context:
```php
            $this->defaultHost = $requestContext->getHost();
            $this->defaultScheme = $requestContext->getScheme();
```
Imho providing those values as parameter in the service definition won't change anything.

alexander-schranz commented:
@stollr as @mamazu mention we should not change the constructor on 2.6.

stollr committed:
Deprecate defining different host and scheme in kernel parameters "router.request_context.*" and "framework.router.default_uri"

The webspace manager, which is configured in the core bundle, relies on the request context container parameters `router.request_context.host` and `router.request_context.scheme` as fallback when no actual request object is available. This might happen during cache warming in a functional test.

But if the request context is configured by Symfony's `framework.router.default_uri` setting like it is described in their [docs](https://symfony.com/doc/current/routing.html#generating-urls-in-commands), the mentioned container parameters still uses the defaults (localhost and http). There is a issue which points this out https://github.com/symfony/symfony/issues/53919.

This leads to wrong URL generation in a functional test (kernel test case) when the cache is warmed.

Therfore a deprecation message will be triggered, now, telling the user that the Symfony config `framework.router.default_uri` needs to have the same host and scheme like the mentioned parameters.

Starting from Sulu 3.0 only the URI defined in `framework.router.default_uri` will be respected.

This change fixes this issue (described in #7280).

stollr commented:
I have simplified the patch. It still changes the constructor a little bit but now, it only checks for difference between `router.request_context.host|schema` and `framework.router.default_uri`. This way users will get a proper deprecation message if they have to change something to prepare for 3.0.

What do you think?

alexander-schranz committed:
Use requestContext in dump sitemap

alexander-schranz committed:
Fix dump sitemap command constructor

alexander-schranz commented:
@stollr Thank you! Ported this to 3.0.
